app/src/main/java/org/mozilla/fenix/home/recentvisits/RecentVisitsFeature.kt

```diff
@@ -0,0 +1,161 @@
+                        val metadataInGroup = group.value
+                        val metadataUrlGroups = metadataInGroup.groupBy { metadata -> metadata.key.url }
+                        metadataUrlGroups.map { metadata ->
+                            metadata.value.reduce { acc, elem ->
+                                acc.copy(
+                                    totalViewTime = acc.totalViewTime + elem.totalViewTime,
+                                    updatedAt = max(acc.updatedAt, elem.updatedAt)
+                                )
+                            }
+                        }
+                    }
+                    .toList()
+            }
+
+            updateState(historyHighlights.await(), recentHistoryGroups.await())
+        }
+    }
+
+    @VisibleForTesting
+    internal fun updateState(
+        historyHighlights: List<HistoryHighlight>,
+        historyGroups: List<Pair<String, List<HistoryMetadata>>>
+    ) {
+        homeStore.dispatch(
+            HomeFragmentAction.RecentHistoryChange(
+                getCombinedHistory(historyHighlights, historyGroups)
+            )
+        )
+    }
+
+    @VisibleForTesting
+    internal fun getCombinedHistory(
+        historyHighlights: List<HistoryHighlight>,
+        historyGroups: List<Pair<String, List<HistoryMetadata>>>
+    ): List<RecentlyVisitedItem> {
+        val dedupedHistoryHighlights = dedupeVisits(historyHighlights, historyGroups)
+        val recentHighlights = dedupedHistoryHighlights
+            .map {
+                RecentHistoryHighlight(
+                    title = if (it.title.isNullOrBlank()) it.url else it.title!!,
+                    url = it.url
+                )
+            }
+        val recentGroups = historyGroups
+            .map { (title, data) ->
+                RecentHistoryGroup(
+                    title = title,
+                    historyMetadata = data
+                )
+            }
+            .sortedByDescending { it.lastUpdated() }
+            .take(MAX_RESULTS_TOTAL)
+
+        val totalItemsCount = recentHighlights.size + recentGroups.size
+
+        return if (totalItemsCount <= MAX_RESULTS_TOTAL) {
+            recentGroups + recentHighlights
+        } else {
+            var groupsCount = 0
+            var highlightCount = 0
+            while ((highlightCount + groupsCount) < MAX_RESULTS_TOTAL) {
+                if (recentGroups.getOrNull(groupsCount) != null) {
+                    groupsCount += 1
+                }
+
+                if ((highlightCount + groupsCount) < MAX_RESULTS_TOTAL &&
+                    recentHighlights.getOrNull(highlightCount) != null
+                ) {
+                    highlightCount += 1
+                }
+            }
+
+            recentGroups.take(groupsCount) + recentHighlights.take(highlightCount)
+        }
+    }
+
+    @VisibleForTesting
+    internal fun dedupeVisits(
```

Comment:
nit: Let's add a comment that we do this to not display a highlight separately that is already in a group.